pkg/providers/aws/provider_redis_metrics.go

```diff
@@ -102,58 +90,100 @@ func (r *RedisMetricsProvider) scrapeRedisCloudWatchMetricData(ctx context.Conte
-	// parse the returned data from the cloudwatch to a GenericCloudMetric
-	var metrics []*providers.GenericCloudMetric
-	for _, metricData := range metricOutput.MetricDataResults {
-		// status code complete ensures all metrics have been successful
-		if *metricData.StatusCode != cloudwatch.StatusCodeComplete {
-			continue
+		// get cluster if for use in metric labels
+		clusterID, err := resources.GetClusterID(ctx, r.Client)
+		if err != nil {
+			return nil, errorUtil.Wrap(err, "error getting clusterID")
 		}
-		// depending on the number of data points, several values can be returned
-		for _, value := range metricData.Values {
-			// convert aws metric data to generic cloud metric data
-			metrics = append(metrics, &providers.GenericCloudMetric{
-				Name: *metricData.Id,
-				Labels: map[string]string{
-					"clusterID":   clusterID,
-					"resourceID":  redis.Name,
-					"namespace":   redis.Namespace,
-					"instanceID":  resourceID,
-					"productName": redis.Labels["productName"],
-					"strategy":    redisProviderName,
-				},
-				Value: *value,
-			})
+
+		// ensure metric data results are not nil
+		if metricOutput.MetricDataResults == nil {
+			return nil, errorUtil.New("no metric data returned from elasticache cloudwatch")
 		}
+
+		logger.Infof("parsing elasticache cloud watch metrics for redis %s", resourceID)
+		// parse the returned data from the cloudwatch to a GenericCloudMetric
+
+		for _, metricData := range metricOutput.MetricDataResults {
+			// status code complete ensures all metrics have been successful
+			if *metricData.StatusCode != cloudwatch.StatusCodeComplete {
+				continue
+			}
+			// depending on the number of data points, several values can be returned
+			for _, value := range metricData.Values {
+				// convert aws metric data to generic cloud metric data
+				metrics = append(metrics, &providers.GenericCloudMetric{
+					Name: *metricData.Id,
+					Labels: map[string]string{
+						"clusterID":   clusterID,
+						"resourceID":  redis.Name,
+						"namespace":   redis.Namespace,
+						"instanceID":  cacheClusterIdString,
+						"productName": redis.Labels["productName"],
+						"strategy":    redisProviderName,
+					},
+					Value: *value,
+				})
+			}
+		}
+
 	}
 	return metrics, nil
 
 }
 
-func buildRedisMetricDataQuery(resourceID string) []*cloudwatch.MetricDataQuery {
-	//todo build this out
-	return nil
+func buildRedisMetricDataQuery(cacheClusterId string, metricTypes []providers.CloudProviderMetricType) []*cloudwatch.MetricDataQuery {
+	var metricDataQueries []*cloudwatch.MetricDataQuery
+	for _, metricType := range metricTypes {
+		metricDataQueries = append(metricDataQueries, &cloudwatch.MetricDataQuery{
+			// id needs to be unique, and is built from the metric name and type, and the CacheClusterId number
+			// the metric name is converted from camel case to snake case to allow it to be easily reused when exposing the metrics
```

Comment:
I think we can remove this comment, it was left over from the original approach

Reply:
agree